### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) and Post-Link Optimization (PLO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [koto-lang/koto#312] [Ideas]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects. The results are available [here](https://github.com/zamazan4ik/awesome-pgo/). According to the tests, PGO can help with achieving better performance in many cases, including compilers and interpreters. Since Koto targets fast compilation (according to the documentation), I think trying to optimize Koto's tooling performance with PGO can be a good idea.

I already did some benchmarks and want to share my results.

## Test environment

* Fedora 39
* Linux kernel 6.8.6
* AMD Ryzen 9 5900x
* 48 Gib RAM
* SSD Samsung 980 Pro 2 Tib
* Compiler - Rustc 1.78-nightly
* koto version: the latest for now from the `main` branch on commit `1856abb432e44c95d40791815771bdc96e1ae34f`
* Disabled Turbo boost

## Benchmark

For benchmark purposes, I use [these](https://github.com/koto-lang/koto/tree/main/koto/benches) benchmarks. For PGO optimization I use [cargo-pgo](https://github.com/Kobzol/cargo-pgo) tool. Release bench result I got with `cargo bench --workspace` command. PGO training phase is done with `cargo pgo bench -- --workspace`, PGO optimization phase - with `cargo pgo optimize bench -- --workspace`.

For all commands `taskset -c 0` is used for reducing the OS scheduler "noise". The measurements are done multiple times - the results are consistent across runs.

## Results

I got the following results:

* Release: https://gist.github.com/zamazan4ik/90521ffd9f25f6f2ba4909bfc93ac302
* PGO optimized compared to Release: https://gist.github.com/zamazan4ik/77a46c1bd726212311358373fba2660e
* (just for reference) PGO instrumentation compared to Release: https://gist.github.com/zamazan4ik/c9361f6caf7cae5a38fa4d57b2ffa358

According to the results, PGO measurably improves Koto's performance at least in the cases above.

## Further steps

I can suggest the following action points:

* Perform more PGO benchmarks with Koto. If it shows improvements - add a note to the documentation about possible improvements in Koto's performance with PGO.
* Providing an easier way (e.g. a build option) to build scripts with PGO can be helpful for the end-users and maintainers since they will be able to optimize Koto tooling according to their workloads.
* Optimize pre-built Koto binaries

Testing Post-Link Optimization techniques (like [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) would be interesting too (Clang and Rustc already use BOLT as an addition to PGO) but I recommend starting from the usual PGO.

Here are some examples of how PGO optimization is integrated into other projects:

* Rustc: a CI [script](https://github.com/rust-lang/rust/blob/master/src/ci/stage-build.py) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback"
  - A [part](https://github.com/gcc-mirror/gcc/blob/4832767db7897be6fb5cbc44f079482c90cb95a6/configure#L7818) in a "wonderful" `configure` script 
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)

I would be happy to answer your questions about PGO and PLO.

**Comment by irh** [comment]
This is wonderful, thanks @zamazan4ik! I didn't know that the PGO results would be quite this impressive, I'll take a closer look during the week.
